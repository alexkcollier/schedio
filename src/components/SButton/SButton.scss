.button {
  @include button-base;

  &--size-small {
    @include button-size-small;
  }

  &--size-regular {
    @include button-size-regular;
  }

  &--size-large {
    @include button-size-large;
  }

  &--loading {
    color: transparent !important;
    pointer-events: none;
    position: relative;
    text-shadow: none !important;
    user-select: none;

    .icon {
      filter: none;
      stroke: transparent !important;
    }

    .button__spinner {
      left: 50%;
      padding-top: spacing('half');
      position: absolute;
      top: 0;
      transform: translateX(-50%);
    }
  }

  &--icon-only {
    padding: spacing('half');

    .icon {
      margin: 0;

      & + .icon {
        margin-left: spacing('half');
      }
    }
  }

  &__icon {
    &--left {
      margin-left: -1 * spacing('quarter');
      margin-right: spacing('tight');
    }

    &--right {
      margin-left: spacing('tight');
      margin-right: -1 * spacing('quarter');
    }

    .button:disabled & {
      filter: none;
      stroke: color('night', 'lightest');
    }
  }

  @each $color-name in $button-colors {
    &--color-#{$color-name} {
      @include button-color($color-name);

      &-outlined {
        @include button-outline($color-name);
      }

      &-text {
        @include button-text($color-name);
      }
    }
  }
}
