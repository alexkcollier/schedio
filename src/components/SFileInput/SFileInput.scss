.file-input {

  // bind parent class name
  $parent: &;

  display: inline-block;

  // hides the default element
  &__control {
    @include control-invisible-base;

    &:focus ~ #{$parent}__display {
      @include control-focused;
    }
  }

  // inner wrapper around control and count
  &__inner {
    align-items: center;
    display: flex;
    font-weight: $font-weight-regular;
    text-transform: none;
  }

  // the element being displayed in place of the <input />
  &__display {
    @include button-base;
    @include button-outline('night');

    &--invalid {
      @include control-invalid;

      border: none !important;
    }
  }

  // count of files
  &__count {
    margin-left: spacing('half');
  }

  // limit of files
  &__limit {
    @include font-size(6);

    color: color('night', 'lighter');
    margin: spacing() spacing('none');
  }

  // list of files
  &__list {
    display: flex;
    flex-wrap: wrap;
    padding: spacing('none');
  }

  // an area with an icon
  &--droppable {
    #{$parent}__inner {
      display: inline-flex;
      flex-direction: column;
      justify-content: center;
    }

    #{$parent}__icon {
      margin: spacing('half');
      margin-top: spacing('none');
      opacity: 0.2;
      pointer-events: none;
      transition: all duration() easing();
    }

    #{$parent}__display {
      $area-width: 12em;

      @include control;

      align-items: center;
      display: flex;
      flex-direction: column;
      justify-content: center;
      padding: spacing('quarter');
      text-align: center;
      width: $area-width;

      &:hover,
      &--dragover {
        background-color: color('ice');
        border-color: color('spartan_blue', 'lighter');

        #{$parent}__icon {
          opacity: 0.8;
        }
      }
    }

    #{$parent}__count {
      margin-left: spacing('none');
      margin-top: spacing('half');
    }
  }

  // disabled state
  &__control:disabled {
    pointer-events: none;

    & + #{$parent}__display {
      @include control-disabled;
    }
  }
}
