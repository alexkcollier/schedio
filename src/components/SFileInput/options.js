export const types = ['button', 'area']
