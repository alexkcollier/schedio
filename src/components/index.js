export { default as SButton } from '@/components/SButton'
export { default as SButtonGroup } from '@/components/SButtonGroup'
export { default as SCard } from '@/components/SCard'
export { default as SCardContainer } from '@/components/SCardContainer'
export { default as SCheckbox } from '@/components/SCheckbox'
export { default as SContainer } from '@/components/SContainer'
export { default as SContainerColumn } from '@/components/SContainerColumn'
export { default as SContainerRow } from '@/components/SContainerRow'
export { default as SHeading } from '@/components/SHeading'
export { default as SIcon } from '@/components/SIcon'
export { default as SInput } from '@/components/SInput'
export { default as SLabel } from '@/components/SLabel'
export { default as SLink } from '@/components/SLink'
export { default as SProgressBar } from '@/components/SProgressBar'
export { default as SRadio } from '@/components/SRadio'
export { default as SSelect } from '@/components/SSelect'
export { default as SSpinner } from '@/components/SSpinner'
export { default as SSwitch } from '@/components/SSwitch'
export { default as STable } from '@/components/STable'
export { default as STableCell } from '@/components/STableCell'
export { default as STableRow } from '@/components/STableRow'
export { default as STextarea } from '@/components/STextarea'
