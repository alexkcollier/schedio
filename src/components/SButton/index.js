import SButton from './SButton.vue';

export { SButton };

export default { install: Vue => Vue.component(SButton.name, SButton) };
