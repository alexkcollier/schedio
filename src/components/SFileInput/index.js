import SFileInput from './SFileInput.vue'

export { SFileInput }

export default {
  install: Vue => Vue.component(SFileInput.name, SFileInput),
}
