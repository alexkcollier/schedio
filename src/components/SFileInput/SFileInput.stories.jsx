import { SFileInput } from '@/components/SFileInput'
import { SCallout } from '@/components/SCallout'
import { SHeading } from '@/components/SHeading'
import PropList from '@/docs/PropList'
import { storiesOf } from '@storybook/vue'
import { boolean, number } from '@storybook/addon-knobs'
import { withAttrsAsProps } from '@/mixins/stories/form-fields'

storiesOf('Components/Inputs.SFileInput', module)
  .addParameters({ jest: 'SFileInput' })
  .add('SFileInput', () => {
    return {
      data() {
        return {
          errors: [],
          files: []
        }
      },

      props: {
        props: {
          default: () => ({
            ...withAttrsAsProps({ id: 'file-input', name: 'file-input', label: 'File input' }),
            maxSize: number('max-size', 0, '', 'Optional Props'),
            isDroppable: boolean('is-droppable', false, 'Optional Props'),
            hideCount: boolean('hide-count', false, 'Optional Props'),
            isInline: boolean('is-inline', false, 'Optional Props')
          })
        },

        attrs: {
          default: {
            disabled: boolean('disabled', false, '$attrs'),
            multiple: boolean('multiple', false, '$attrs')
          }
        }
      },
      render(h) {
        const { props, attrs } = this.$props

        return (
          <div>
            <SHeading level="1">SFileInput</SHeading>
            <p>
              This component allows uses to submit files. The <code>is-droppable</code> prop enables
              drag-and-drop functionality.
            </p>
            <p>
              <code>SFileInput</code> can be used with <code>v-model</code>. It will emit an
              <code>input</code> event with an array of files as its payload.
            </p>

            <SHeading level="2">Example</SHeading>
            <SFileInput
              {...{ props }}
              {...{ attrs }}
              onError={evt => (this.errors = evt.payload)}
              onInput={evt => (this.files = evt)}
              files={this.files}
            />

            {!!this.errors.length && (
              <div>
                <SFormValidation>
                  Example error handling
                  <pre class="text--color-red">{JSON.stringify(this.errors, null, 2)}</pre>
                </SFormValidation>
                <p>
                  Error handler: <code>evt => (this.errors = evt.payload)</code>
                </p>
              </div>
            )}

            <SCallout>
              Note that the <code>max-size</code> prop is specified in bytes.
            </SCallout>

            <SHeading level="2">Validation</SHeading>
            <p>
              <code>SFileInput</code> provides limited validation for maximum file size and
              duplicate file names, but no more. On a validation error, it emits an{' '}
              <code>error</code> event, and you have to handle reporting the error to the user
              yourself. The event looks like this:
            </p>
            <pre>
              {`\
{
  payload: [
    {
      name: 'Failed file',
      reasons: ['exceeds max size', 'duplicate name']
    },
    ...moreFiles
  ]
}`}
            </pre>

            {SFileInput.props && <PropList component={SFileInput} />}
          </div>
        )
      }
    }
  })
