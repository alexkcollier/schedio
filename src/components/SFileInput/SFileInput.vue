<template>
  <div
    class="file-input"
    :class="classList"
  >
    <SLabel
      :for="id"
      :is-required="required"
      :hide-optional="hideOptional"
      :label="label"
    >
      <div class="file-input__inner">
        <!-- input receives non-drag events -->
        <input
          :id="id"
          v-bind="$attrs"
          :label="label"
          :name="name"
          type="file"
          class="file-input__control"
          v-on="listeners.nondragListeners"
        >

        <!-- only handles drag listeners when component is not a button -->
        <div
          class="file-input__display"
          :class="{
            'file-input__display--invalid': isInvalid,
            'file-input__display--dragover': hasDragover,
          }"
          v-on="listeners.dragListeners"
        >
          <SIcon
            v-if="!hideIcon && isDroppable"
            icon="upload-cloud"
            size="large"
            class="file-input__icon"
          />

          {{ displayText }}
        </div>

        <span
          v-if="!$attrs.disabled && !hideCount"
          class="file-input__count"
        >{{ fileCountText }}</span>
      </div>
    </SLabel>

    <output class="file-input__info">
      <p
        v-if="maxSize"
        class="file-input__limit"
      >
        Maximum file size: {{ maxSize }} bytes
      </p>

      <ul
        v-show="fileList.length"
        class="file-input__list"
      >
        <SChip
          v-for="(file, idx) in fileList"
          :key="file.name"
          tag="li"
          is-closable
          @close="removeFile({ idx })"
        >
          {{ file.name }}
        </SChip>
      </ul>
    </output>
  </div>
</template>

<script>
import InputProps from '@/mixins/InputProps.mixin'
import { SLabel } from '@/components/SLabel'
import { SChip } from '@/components/SChip'

export default {
  name: 'SFileInput',

  components: {
    SLabel,
    SChip
  },

  mixins: [InputProps],

  inheritAttrs: false,

  model: {
    event: 'input',
    prop: 'files'
  },

  props: {
    maxSize: {
      default: null,
      type: [Number, String]
    },

    text: {
      default: '',
      type: String
    },

    isDroppable: {
      default: false,
      type: Boolean
    },

    hideIcon: {
      default: false,
      type: Boolean
    },

    hideCount: {
      default: false,
      type: Boolean
    },

    files: {
      default: () => [],
      type: Array
    }
  },

  data() {
    return {
      fileList: [],
      hasDragover: false
    }
  },

  computed: {
    fileNames() {
      return this.fileList.map(file => file.name)
    },

    displayText() {
      const defaultText = this.$attrs.multiple ? 'Choose files' : 'Choose a file'

      if (this.isDroppable) return `Drag and drop or click here to ${defaultText.toLowerCase()}`

      return this.text || defaultText
    },

    fileCountText() {
      const count = this.fileList.length
      let text = `${count} file`

      if (count !== 1) text += 's'

      return `${text} selected`
    },

    classList() {
      return {
        'file-input--droppable': this.isDroppable,
        'file-input--multipe': this.$attrs.multiple
      }
    },

    listeners() {
      const { dragenter, dragover, dragleave, drop, ...otherListeners } = this.$listeners
      let dragListeners = {}

      if (this.isDroppable) {
        dragListeners = {
          dragenter: evt => this.handleDragenter(evt),
          dragover: evt => this.handleDragover(evt),
          dragleave: evt => this.handleDragleave(evt),
          drop: evt => this.handleDrop(evt)
        }
      }

      const nondragListeners = {
        ...otherListeners,
        input: evt => this.handleFiles(evt)
      }

      return { nondragListeners, dragListeners }
    }
  },

  watch: {
    files(val) {
      this.fileList = val
    },

    fileList(val) {
      if (val && Array.isArray(val) && val.length) this.$emit('input', this.fileList)
    }
  },

  methods: {
    handleFiles(event) {
      const eventItem = event.dataTransfer ? event.dataTransfer : event.target
      const filesArray = eventItem.files.length ? Array.from(eventItem.files) : []
      const { invalid, valid } = this.getValidFiles(filesArray)

      if (invalid.length > 0) {
        const payload = invalid.map(({ name, failed }) => ({ name, failed }), [])
        this.$emit('error', { payload })
      }

      this.fileList = this.$attrs.multiple ? [...this.fileList, ...valid] : valid

      // reset the input value
      event.target.value = null
    },
    /**
     * Validates files
     * @param {Array.<File>} files - The array of files
     *
     * @return {{ invalid: File[], valid: File[] }} - An object of valid and invalid files
     */
    getValidFiles(files) {
      const validate = ({ invalid, valid }, file) => {
        const { isValid, failed } = this.checkValidations(file)

        if (isValid) {
          valid.push(file)
        } else {
          file.failed = failed
          invalid.push(file)
        }

        return { invalid, valid }
      }

      return files.reduce(validate, { invalid: [], valid: [] })
    },

    checkValidations(file) {
      const toValidate = [
        {
          type: 'exceeds max size',
          isValid: this.isValidSize(file)
        },
        {
          type: 'duplicate name',
          isValid: this.hasUniqueName(file)
        }
      ]

      const failed = toValidate.reduce((r, { isValid, type }) => (isValid ? r : r.concat(type)), [])

      return { isValid: failed.length === 0, failed }
    },

    isValidSize(file) {
      return this.maxSize ? file.size <= this.maxSize : true
    },

    hasUniqueName(file) {
      return this.$attrs.multiple ? !this.fileNames.includes(file.name) : true
    },

    handleDragenter(event) {
      this.stopAndPrevent(event)
      this.hasDragover = true
    },

    handleDragleave(event) {
      this.stopAndPrevent(event)
      this.hasDragover = false
    },

    handleDrop(event) {
      this.stopAndPrevent(event)
      this.hasDragover = false
      this.handleFiles(event)
    },

    handleDragover(event) {
      this.stopAndPrevent(event)
    },

    stopAndPrevent(event) {
      event.stopPropagation()
      event.preventDefault()
    },

    removeFile({ idx }) {
      this.fileList.splice(idx, 1)
    }
  }
}
</script>
