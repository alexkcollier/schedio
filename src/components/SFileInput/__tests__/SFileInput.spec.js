import { shallowMount, mount } from '@vue/test-utils'
import SFileInput from '../SFileInput.vue'
import MockFile from '../../../../tests/__mocks__/MockFile'
import { SChip } from '@/components/SChip'
import { SButton } from '@/components/SButton'
import { axe, toHaveNoViolations } from 'jest-axe'

expect.extend(toHaveNoViolations)

describe('SFileInput.vue', () => {
  let wrapper
  const errorSpy = jest.spyOn(global.console, 'error').mockImplementation(() => {})

  // mock files
  const fileA = new MockFile({
    name: 'test-file-a.txt',
    size: 44320,
    type: 'text/plain'
  })

  const fileB = new MockFile({
    name: 'test-file-b.txt',
    size: 512,
    type: 'text/plain'
  })

  const defaultProps = {
    id: 'file-input',
    name: 'file-input',
    label: 'File input'
  }

  beforeEach(() => {
    wrapper = shallowMount(SFileInput, {
      propsData: defaultProps
    })
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('renders correctly', async () => {
    wrapper = mount(SFileInput, { propsData: defaultProps })
    expect(wrapper.html()).toMatchSnapshot()
    expect(await axe(wrapper.html())).toHaveNoViolations()
  })

  it('requires id, name, and label', () => {
    shallowMount(SFileInput)
    // 4 times because we one of the props expects a string, but receives `undefined`
    expect(errorSpy).toBeCalledTimes(4)
    expect(errorSpy.mock.calls[0][0]).toContain('[Vue warn]: Missing required prop')
  })

  /**
   * We can't test the upload functionality with Jest currently.
   * jsdom is missing the `DataTransfer` object.
   * Instead, we're just going to test the input's input handler.
   */
  it('handles inputs', () => {
    // watch the input handler
    const inputSpy = jest.spyOn(wrapper.vm, 'handleFiles')
    wrapper.setMethods({ handleFiles: inputSpy })
    // ensure input is triggered and handler is called
    wrapper.find('input').trigger('input')
    expect(inputSpy).toBeCalled()
    // fake the input event
    wrapper.vm.handleFiles({ target: { files: [fileA] } })
    expect(wrapper.emitted('input')).toBeTruthy()
    expect(wrapper.contains(SChip)).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('passes `multiple` to the input', async () => {
    wrapper = mount(SFileInput, { propsData: { ...defaultProps, multiple: true } })
    expect(wrapper.find('input').element.multiple).toBeTruthy()
    expect(await axe(wrapper.html())).toHaveNoViolations()
  })

  it('can upload multiple files', () => {
    wrapper.vm.handleFiles({ target: { files: [fileA, fileB] } })
    expect(wrapper.vm.fileList.length).toBe(2)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can remove files', () => {
    wrapper = mount(SFileInput, {
      propsData: defaultProps
    })

    // emitted[0]
    wrapper.vm.handleFiles({ target: { files: [fileA] } })
    expect(wrapper.vm.fileList.length).toBe(1)
    expect(wrapper.emitted('input')).toBeTruthy()
    expect(wrapper.emitted('input')[0][0]).toEqual(wrapper.vm.fileList)

    // find the button and click it
    // emitted[1] - `click` also triggers an input event
    wrapper.find(SButton).trigger('click')
    expect(wrapper.vm.fileList.length).toBe(0)
    expect(wrapper.emitted('input')[1][0]).toEqual(wrapper.vm.fileList)

    // with multiple files
    // emitted[0]
    wrapper.vm.handleFiles({ target: { files: [fileA, fileB] } })
    expect(wrapper.vm.fileList.length).toBe(2)
    expect(wrapper.emitted('input')[2][0]).toEqual(wrapper.vm.fileList)

    // find all buttons, click the first
    // emitted[3]
    wrapper
      .findAll(SButton)
      .at(0)
      .trigger('click')
    expect(wrapper.vm.fileList.length).toBe(1)
    expect(wrapper.emitted('input')[3][0]).toEqual(wrapper.vm.fileList)

    // ensure only the desired file was deleted
    expect(wrapper.vm.fileNames).toContain(fileB.name)
  })

  it('works with v-model', () => {
    wrapper = mount({
      components: { SFileInput },
      data: () => ({ files: [], defaultProps }),
      template: '<SFileInput v-bind="defaultProps" v-model="files" />'
    })

    const fileInput = wrapper.find(SFileInput)

    fileInput.vm.handleFiles({ target: { files: [fileA] } })
    expect(wrapper.vm.files).toEqual(fileInput.vm.fileList)
  })

  it('displays a file count', () => {
    wrapper.setProps({ multiple: true })
    wrapper.vm.handleFiles({ target: { files: [fileA, fileB] } })
    expect(wrapper.text()).toContain('2 file')
  })

  it('can hide a file count', () => {
    wrapper.setProps({ multiple: true, hideCount: true })
    wrapper.vm.handleFiles({ target: { files: [fileA, fileB] } })
    expect(wrapper.text()).not.toContain('2 file')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('checks for duplicate file names', () => {
    wrapper.setProps({ multiple: true })
    wrapper.vm.handleFiles({ target: { files: [fileA, fileB] } })
    wrapper.vm.handleFiles({ target: { files: [fileA] } })

    const emittedError = wrapper.emitted('error')
    expect(emittedError).toBeTruthy()
    expect(emittedError[0][0].payload[0].failed).toContain('duplicate name')
    expect(emittedError[0][0].payload[0].name).toBe(fileA.name)
    expect(wrapper.vm.fileList.length).toBe(2)
  })

  it('displays file names to users', () => {
    wrapper.vm.handleFiles({ target: { files: [fileA, fileB] } })
    expect(wrapper.vm.fileNames).toEqual([fileA, fileB].map(file => file.name))
    expect(wrapper.text()).toContain(fileA.name)
    expect(wrapper.text()).toContain(fileB.name)
  })

  it('hides the file list if no files exist', () => {
    expect(wrapper.find('ul').isVisible()).toBe(false)
  })

  it('checks for max file size', () => {
    wrapper.setProps({ multiple: true, maxSize: 1024 })
    wrapper.vm.handleFiles({ target: { files: [fileA, fileB] } })

    const emittedError = wrapper.emitted('error')
    expect(emittedError).toBeTruthy()
    expect(emittedError[0][0].payload[0].failed).toContain('exceeds max size')
    expect(emittedError[0][0].payload[0].name).toBe(fileA.name)
    expect(wrapper.vm.fileList.length).toBe(1)
  })

  it('displays the file size limit', () => {
    const maxSize = 23829
    wrapper.setProps({ maxSize })
    expect(wrapper.text()).toContain(maxSize)
  })

  it('can have custom text', () => {
    const customText = 'Custom text'
    wrapper.setProps({ text: customText })
    expect(wrapper.text()).toContain(customText)
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('drop zone', () => {
    let display

    beforeEach(() => {
      wrapper.setProps({ isDroppable: true })
      display = wrapper.find('.file-input__display')
    })

    it('can be a drop zone', () => {
      expect(errorSpy).not.toBeCalled()
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('can hide the icon', () => {
      wrapper.setProps({ hideIcon: true })
      expect(wrapper.contains('sicon')).toBe(false)
      expect(wrapper.html()).toMatchSnapshot()
    })

    it.each`
      event          | handlers                         | includesClass                      | excludesClass
      ${'dragenter'} | ${['handleDragenter']}           | ${'file-input__display--dragover'} | ${''}
      ${'dragover'}  | ${['handleDragover']}            | ${''}                              | ${''}
      ${'dragleave'} | ${['handleDragleave']}           | ${''}                              | ${'file-input__display--dragover'}
      ${'drop'}      | ${['handleDrop', 'handleFiles']} | ${''}                              | ${''}
    `('handles $event correctly', ({ event, handlers, includesClass, excludesClass }) => {
      const spies = new Map(handlers.map(handler => [handler, jest.spyOn(wrapper.vm, handler)]))

      spies.forEach((spy, handler) => wrapper.setMethods({ [handler]: spy }))
      display.trigger(event)

      spies.forEach(spy => expect(spy).toBeCalled())
      if (includesClass) expect(display.classes()).toContain(includesClass)
      if (excludesClass) expect(display.classes()).not.toContain(excludesClass)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
