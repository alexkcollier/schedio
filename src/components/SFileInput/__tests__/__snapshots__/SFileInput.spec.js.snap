// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SFileInput.vue can have custom text 1`] = `
<div class="file-input">
  <slabel-stub label="File input" isrequired="true" for="file-input">
    <div class="file-input__inner"><input id="file-input" label="File input" name="file-input" type="file" class="file-input__control">
      <div class="file-input__display">
        <!---->

        Custom text
      </div> <span class="file-input__count">0 files selected</span>
    </div>
  </slabel-stub> <output class="file-input__info">
    <!---->
    <schipgroup-stub color="" tag="ul" class="file-input__list" style="display: none;"></schipgroup-stub>
  </output>
</div>
`;

exports[`SFileInput.vue can hide a file count 1`] = `
<div class="file-input">
  <slabel-stub label="File input" isrequired="true" for="file-input">
    <div class="file-input__inner"><input id="file-input" label="File input" name="file-input" type="file" class="file-input__control" multiple="multiple">
      <div class="file-input__display">
        <!---->

        Choose a file
      </div>
      <!---->
    </div>
  </slabel-stub> <output class="file-input__info">
    <!---->
    <schipgroup-stub color="" tag="ul" class="file-input__list" style="">
      <schip-stub color="" isclosable="true" isactive="true" tag="li" closearialabel="Remove test-file-a.txt">
        test-file-a.txt
      </schip-stub>
      <schip-stub color="" isclosable="true" isactive="true" tag="li" closearialabel="Remove test-file-b.txt">
        test-file-b.txt
      </schip-stub>
    </schipgroup-stub>
  </output>
</div>
`;

exports[`SFileInput.vue can upload multiple files 1`] = `
<div class="file-input">
  <slabel-stub label="File input" isrequired="true" for="file-input">
    <div class="file-input__inner"><input id="file-input" label="File input" name="file-input" type="file" class="file-input__control">
      <div class="file-input__display">
        <!---->

        Choose a file
      </div> <span class="file-input__count">2 files selected</span>
    </div>
  </slabel-stub> <output class="file-input__info">
    <!---->
    <schipgroup-stub color="" tag="ul" class="file-input__list" style="">
      <schip-stub color="" isclosable="true" isactive="true" tag="li" closearialabel="Remove test-file-a.txt">
        test-file-a.txt
      </schip-stub>
      <schip-stub color="" isclosable="true" isactive="true" tag="li" closearialabel="Remove test-file-b.txt">
        test-file-b.txt
      </schip-stub>
    </schipgroup-stub>
  </output>
</div>
`;

exports[`SFileInput.vue drop zone can be a drop zone 1`] = `
<div class="file-input file-input--droppable">
  <slabel-stub label="File input" isrequired="true" for="file-input">
    <div class="file-input__inner"><input id="file-input" label="File input" name="file-input" type="file" class="file-input__control">
      <div class="file-input__display">
        <sicon-stub icon="upload-cloud" color="" size="large" class="file-input__icon"></sicon-stub>

        Drag and drop or click here to choose a file
      </div> <span class="file-input__count">0 files selected</span>
    </div>
  </slabel-stub> <output class="file-input__info">
    <!---->
    <schipgroup-stub color="" tag="ul" class="file-input__list" style="display: none;"></schipgroup-stub>
  </output>
</div>
`;

exports[`SFileInput.vue drop zone can hide the icon 1`] = `
<div class="file-input file-input--droppable">
  <slabel-stub label="File input" isrequired="true" for="file-input">
    <div class="file-input__inner"><input id="file-input" label="File input" name="file-input" type="file" class="file-input__control">
      <div class="file-input__display">
        <!---->

        Drag and drop or click here to choose a file
      </div> <span class="file-input__count">0 files selected</span>
    </div>
  </slabel-stub> <output class="file-input__info">
    <!---->
    <schipgroup-stub color="" tag="ul" class="file-input__list" style="display: none;"></schipgroup-stub>
  </output>
</div>
`;

exports[`SFileInput.vue drop zone handles "dragenter" correctly 1`] = `
<div class="file-input file-input--droppable">
  <slabel-stub label="File input" isrequired="true" for="file-input">
    <div class="file-input__inner"><input id="file-input" label="File input" name="file-input" type="file" class="file-input__control">
      <div class="file-input__display file-input__display--dragover">
        <sicon-stub icon="upload-cloud" color="" size="large" class="file-input__icon"></sicon-stub>

        Drag and drop or click here to choose a file
      </div> <span class="file-input__count">0 files selected</span>
    </div>
  </slabel-stub> <output class="file-input__info">
    <!---->
    <schipgroup-stub color="" tag="ul" class="file-input__list" style="display: none;"></schipgroup-stub>
  </output>
</div>
`;

exports[`SFileInput.vue drop zone handles "dragleave" correctly 1`] = `
<div class="file-input file-input--droppable">
  <slabel-stub label="File input" isrequired="true" for="file-input">
    <div class="file-input__inner"><input id="file-input" label="File input" name="file-input" type="file" class="file-input__control">
      <div class="file-input__display">
        <sicon-stub icon="upload-cloud" color="" size="large" class="file-input__icon"></sicon-stub>

        Drag and drop or click here to choose a file
      </div> <span class="file-input__count">0 files selected</span>
    </div>
  </slabel-stub> <output class="file-input__info">
    <!---->
    <schipgroup-stub color="" tag="ul" class="file-input__list" style="display: none;"></schipgroup-stub>
  </output>
</div>
`;

exports[`SFileInput.vue drop zone handles "dragover" correctly 1`] = `
<div class="file-input file-input--droppable">
  <slabel-stub label="File input" isrequired="true" for="file-input">
    <div class="file-input__inner"><input id="file-input" label="File input" name="file-input" type="file" class="file-input__control">
      <div class="file-input__display">
        <sicon-stub icon="upload-cloud" color="" size="large" class="file-input__icon"></sicon-stub>

        Drag and drop or click here to choose a file
      </div> <span class="file-input__count">0 files selected</span>
    </div>
  </slabel-stub> <output class="file-input__info">
    <!---->
    <schipgroup-stub color="" tag="ul" class="file-input__list" style="display: none;"></schipgroup-stub>
  </output>
</div>
`;

exports[`SFileInput.vue drop zone handles "drop" correctly 1`] = `
<div class="file-input file-input--droppable">
  <slabel-stub label="File input" isrequired="true" for="file-input">
    <div class="file-input__inner"><input id="file-input" label="File input" name="file-input" type="file" class="file-input__control">
      <div class="file-input__display">
        <sicon-stub icon="upload-cloud" color="" size="large" class="file-input__icon"></sicon-stub>

        Drag and drop or click here to choose a file
      </div> <span class="file-input__count">0 files selected</span>
    </div>
  </slabel-stub> <output class="file-input__info">
    <!---->
    <schipgroup-stub color="" tag="ul" class="file-input__list" style="display: none;"></schipgroup-stub>
  </output>
</div>
`;

exports[`SFileInput.vue handles inputs 1`] = `
<div class="file-input">
  <slabel-stub label="File input" isrequired="true" for="file-input">
    <div class="file-input__inner"><input id="file-input" label="File input" name="file-input" type="file" class="file-input__control">
      <div class="file-input__display">
        <!---->

        Choose a file
      </div> <span class="file-input__count">1 file selected</span>
    </div>
  </slabel-stub> <output class="file-input__info">
    <!---->
    <schipgroup-stub color="" tag="ul" class="file-input__list" style="">
      <schip-stub color="" isclosable="true" isactive="true" tag="li" closearialabel="Remove test-file-a.txt">
        test-file-a.txt
      </schip-stub>
    </schipgroup-stub>
  </output>
</div>
`;

exports[`SFileInput.vue renders correctly 1`] = `
<div class="file-input"><label class="label" for="file-input">
    <div>
      File input
      <!---->
    </div>
    <div class="file-input__inner"><input id="file-input" label="File input" name="file-input" type="file" class="file-input__control">
      <div class="file-input__display">
        <!---->

        Choose a file
      </div> <span class="file-input__count">0 files selected</span>
    </div>
  </label> <output class="file-input__info">
    <!---->
    <ul class="chip-group file-input__list" style="display: none;"></ul>
  </output></div>
`;
